src/Sentry.Unity/Integrations/ANRIntegration.cs

```diff
@@ -0,0 +1,134 @@
+        }
+
+        internal void StartOnce(IDiagnosticLogger? logger, SentryMonoBehaviour monoBehaviour)
+        {
+            // Start the thread, if not yet running.
+            lock (_cancel)
+            {
+                if (_thread is null)
+                {
+                    _monoBehaviour = monoBehaviour;
+                    _logger = logger;
+                    _thread = new Thread(Run)
+                    {
+                        Name = "Sentry-ANR-WatchDog"
+                    };
+                    _thread.Start();
+
+                    // Stop the thread when the app is being shut down.
+                    _monoBehaviour.Application.Quitting += () =>
+                    {
+                        _cancel.Cancel();
+                        _thread.Join();
+                    };
+
+                    // Update the UI status periodically by running a couroutine on the UI thread
+                    _monoBehaviour.StartCoroutine(UpdateUiStatus());
+                }
+            }
+        }
+
+        private IEnumerator UpdateUiStatus()
+        {
+            var waitForSeconds = new WaitForSeconds((float)Math.Min(sleepInterval, _checkIntervalMilliseconds) / 1000);
+
+            yield return waitForSeconds;
+            while (!_cancel.IsCancellationRequested)
+            {
+                _uiIsResponsive = true;
+                _reported = false;
+                yield return waitForSeconds;
+            }
+        }
+
+        private void Run()
+        {
+            try
+            {
+                long msUntilCheck; // avoiding allocs in the loop
+                var watch = Stopwatch.StartNew();
+
+                while (!_cancel.IsCancellationRequested)
+                {
+                    msUntilCheck = _checkIntervalMilliseconds - watch.ElapsedMilliseconds;
+                    if (msUntilCheck > 0)
+                    {
+                        // sleep for a bit
+                        Thread.Sleep(msUntilCheck < sleepInterval ? (int)msUntilCheck : sleepInterval);
+                    }
+                    else
+                    {
+                        // time to check
+                        if (!_uiIsResponsive && !_reported)
+                        {
+                            var message = $"Application not responding for at least {_checkIntervalMilliseconds} ms.";
+                            _logger?.LogInfo("Detected an ANR event: {0}", message);
+                            OnApplicationNotResponding?.Invoke(this, new ApplicationNotResponding(message));
+                            _reported = true;
+                        }
+
+                        _uiIsResponsive = false;
+                        watch.Restart();
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                _logger?.Log(SentryLevel.Warning, "Exception in the ANR watchdog.", e);
```

Comment:
Perhaps best to use Level Error?

Reply:
added a catch clause for ThreadAbortException with a warning and changed this to error